### ISSUE TITLE
Unable to discriminate between self-referential types when using unions

### ISSUE THREAD

**Issue by brearleyt** [microsoft/pyright#6451]
**Describe the bug**
My type-fu is weak, but it seems that we are currently unable to use `isinstance()` to discriminate between self-referential types when used via a union. 

**Code or Screenshots**

```python
import typing as t

Foo = list["Bar"] | dict[str, "Bar"]
Bar = str | Foo


def passing_function(foo: Foo) -> None:
    t.reveal_type(foo)
    if isinstance(foo, dict):
        t.reveal_type(foo)
        foo["foo"] = "bar"


def failing_function(bar: Bar) -> None:
    t.reveal_type(bar)
    if isinstance(bar, dict):
        t.reveal_type(bar)
        bar["foo"] = "bar"
```

Output:
```
    example.py
      example.py:8:19 - information: Type of "foo" is "list[str | Foo] | dict[str, str | Foo]"
      example.py:10:23 - information: Type of "foo" is "dict[str, str | list[Bar] | ...]"
      example.py:15:19 - information: Type of "bar" is "str | list[Bar] | dict[str, Bar]"
      example.py:17:23 - information: Type of "bar" is "list[str | Foo] | dict[str, str | Foo]"
      example.py:18:9 - error: No overloads for "__setitem__" match the provided arguments (reportGeneralTypeIssues)
      example.py:18:9 - error: Argument of type "Literal['foo']" cannot be assigned to parameter "__key" of type "slice" in function "__setitem__"
        "Literal['foo']" is incompatible with "slice" (reportGeneralTypeIssues)
```

**VS Code extension or command-line**
Running pyright v.1.1.336 via the command line

**Comment by erictraut** [comment]
Thanks for the bug report. I'm able to repro the problem and agree it looks like a bug. If you swap `Bar` for `str | Foo`, the problem goes away, so it appears to have something to do with a type alias that references another recursive type alias. I'll investigate further.

**Comment by erictraut** [comment]
It was an easy fix, and it will be addressed in the next release. I've added a variant of your code sample to the pyright unit tests to prevent future regressions.

**Comment by brearleyt** [comment]
Great, thanks for the rapid response @erictraut.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.337, which I just published. It will also be included in a future release of pylance.
